java/arcs/core/analysis/DependencyNode.kt

```diff
@@ -0,0 +1,131 @@
+ *
+ * Example:
+ *   ```
+ *   particle HousePets
+ *     input: reads PetData { cats: Number, dogs: Number, household: Text }
+ *     output: writes PetStats {
+ *       ratio: inline Domestication {trained: Number, total: Number},
+ *       family: Text,
+ *       limit: Number
+ *     } = new PetStats {
+ *       ratio: new Domestication {
+ *         trained: input.cats,
+ *         total: input.cats + input.dogs,
+ *       },
+ *       family: input.household,
+ *       limit: 5
+ *     }
+ *   ```
+ *
+ *   This can be translated to:
+ *
+ *   ```
+ *   DependencyNode.AggregateValue(
+ *    "ratio" to DependencyNode.AggregateValue(
+ *      "trained" to DependencyNode.PrimitiveValue("input", "cats"),
+ *      "total" to DependencyNode.DerivedFrom(
+ *        DependencyNode.PrimitiveValue("input", "cats"),
+ *        DependencyNode.PrimitiveValue("input", "dogs")
+ *      )
+ *    ),
+ *    "family" to DependencyNode.PrimitiveValue("input", "household"),
+ *    "limit" to DependencyNode.LITERAL
+ *   )
+ *   ```
+ *
+ *   We can represent this DAG graphically:
+ *
+ *                    (root)
+ *            __________|______
+ *           /      |          \
+ *          /      /        _(ratio)_
+ *        /       /        |         \
+ *   (limit)  (family)  (trained)  (total)
+ *     |        |          |     /   |
+ *     |        |          |   /     |
+ *     x     (household)  (cats)  (dogs)
+ *
+ *   This internal representation, in turn, can be translated into the following claims:
+ *
+ *   ```
+ *   claim output.ratio.trained derives from input.cats
+ *   claim output.ratio.total derives from input.cats and derives from input.dogs
+ *   claim output.family derives from input.household
+ *   ```
+ */
+sealed class DependencyNode {
+
+  /** An unmodified input (from a handle connection) used in a Paxel [Expression]. */
+  data class PrimitiveValue(val path: Path = emptyList()) : DependencyNode() {
+    constructor(vararg fields: Identifier) : this(listOf(*fields))
+  }
+
+  /** Represents modification of an access path within a Paxel [Expression]. */
+  data class DerivedFrom(val inputs: Set<DependencyNode> = emptySet()) : DependencyNode() {
+
+    constructor() : this(emptySet())
+
+    constructor(vararg paths: Path) : this(paths.map { PrimitiveValue(it) }.toSet())
+
+    constructor(vararg nodes: DependencyNode) : this(flatten(*nodes))
+
+    companion object {
+      /** Flatten nested sets of [DependencyNode]s.*/
+      private fun flatten(vararg nodes: DependencyNode): Set<DependencyNode> {
+        return nodes.flatMap { node ->
+          when (node) {
+            is PrimitiveValue -> setOf(node)
+            is DerivedFrom -> node.inputs
+            else -> throw IllegalArgumentException(
+              "Nodes must be a 'PrimitiveValue' or 'DerivedFrom'."
+            )
+          }
+        }.toSet()
+      }
+    }
+  }
+
+  /** Associates [Identifier]s with [DependencyNode]s. */
+  data class AggregateValue(
```

Comment:
It seems to me that this should not be a `DependencyNode`. Rather, we should call it `PaxelScope` (or something similar) and the visitor should work on a`Pair<PaxelScope?, DependencyNode>` or an equivalent data class.

Reply:
Right now, it's a dependency node so we can nest scopes. If it's not a dependency node, what's a good way to accomplish this?

Reply:
Many cycles ago, I had an implementation that's similar to what you're proposing. I had a tuple of Paths and Scopes. We moved away from this approach because if created a complicated implementation and it was hard to cover all the cases. Back then, I didn't explore use of visitor contexts, so it's possible that going down that path now would prove successful. However, right now, we're near a clean, working implementation, and investigating this seems like it wouldn't buy us that much more.

Reply:
Yes, I realized the problem with nestedness after I submitted the review. It would also changes the tests substantially. Let us leave it as is for now.